aten/src/ATen/SparseTensorImpl.h

```diff
@@ -64,41 +56,70 @@ struct AT_API SparseTensorImpl : public TensorImpl {
   void * unsafeGetTH(bool retain) override;
   std::unique_ptr<Storage> storage() override;
 
-  // Some ops do some manual size fiddling.
-  // TODO: Figure out a more safe way to provide this functionality
-  std::vector<int64_t>& _sizes_mut() { return size_; }
-
   // WARNING: This function does NOT preserve invariants of sparseDims/denseDims with
   // respect to indices and values
   void raw_resize_(int64_t sparseDims, int64_t denseDims, ArrayRef<int64_t> size) {
-    // UGHHHHH.  Legacy special case
-    if (size.size() == 0) {
-      size_ = {0};
-    } else {
-      size_ = size.vec();
-    }
+    size_ = size.vec();
     sparseDims_ = sparseDims;
     denseDims_ = denseDims;
   }
 
-  // TODO: I hate these two setters, please get rid of them!!!
-  void set_indices(const Tensor& indices) {
-    AT_ASSERT(indices.type().backend() == at::toDense(type().backend()));
-    AT_ASSERT(indices.type().scalarType() == kLong);
-    indices_ = indices;
+  // NOTE: This function preserves invariants of sparseDims/denseDims with respect to
+  // indices and values.
+  void resize_(int64_t sparseDims, int64_t denseDims, ArrayRef<int64_t> size) {
+    AT_CHECK(sparseDims + denseDims == size.size(), "number of dimensions must be sparseDims (", sparseDims, ") + denseDims (", denseDims, "), but got ", size.size());
+
+    if ((!size.equals(size_)) || (sparseDims != sparseDims_) || (denseDims != denseDims_)) {
+      std::vector<int64_t> values_size = {values().size(0)};
+      auto dense_size = size.slice(sparseDims);
+      values_size.insert(values_size.end(), dense_size.begin(), dense_size.end());
+      values_.resize_(values_size);
+
+      std::vector<int64_t> indices_size = indices().sizes().vec();
+      indices_size[0] = sparseDims;
+      indices_.resize_(indices_size);
+    }
```

Comment:
For `sparseDims != sparseDims_` case, we should throw an error saying that this does not preserve data and thus is not allowed.

Reply:
Update: fixed.